File: frontend/src/charts/indicators.ts
/** Hilega Milega indicators — mirrors backend app/engine/indicators.py */

export interface IndicatorParams {
  rsiLength: number
  wmaLength: number
  emaLength: number
  midLevel: number
}

export const DEFAULT_INDICATOR_PARAMS: IndicatorParams = {
  rsiLength: 14,
  wmaLength: 21,
  emaLength: 3,
  midLevel: 50,
}

export function computeRsi(closes: number[], length: number): number[] {
  const rsi = Array(closes.length).fill(NaN)
  if (closes.length < length + 1) return rsi

  const deltas = closes.slice(1).map((c, i) => c - closes[i])
  let avgGain = 0
  let avgLoss = 0
  for (let i = 0; i < length; i++) {
    const d = deltas[i]
    if (d > 0) avgGain += d
    else avgLoss += -d
  }
  avgGain /= length
  avgLoss /= length

  rsi[length] = avgLoss === 0 ? 100 : 100 - 100 / (1 + avgGain / avgLoss)

  for (let i = length + 1; i < closes.length; i++) {
    const d = deltas[i - 1]
    avgGain = (avgGain * (length - 1) + Math.max(d, 0)) / length
    avgLoss = (avgLoss * (length - 1) + Math.max(-d, 0)) / length
    rsi[i] = avgLoss === 0 ? 100 : 100 - 100 / (1 + avgGain / avgLoss)
  }
  return rsi
}

export function computeWma(values: number[], length: number): number[] {
  const out = Array(values.length).fill(NaN)
  const weights = Array.from({ length }, (_, i) => i + 1)
  const wSum = weights.reduce((a, b) => a + b, 0)
  for (let i = length - 1; i < values.length; i++) {
    const window = values.slice(i - length + 1, i + 1)
    if (window.some((v) => Number.isNaN(v))) continue
    out[i] = window.reduce((sum, v, j) => sum + v * weights[j], 0) / wSum
  }
  return out
}

export function computeEma(values: number[], length: number): number[] {
  const out = Array(values.length).fill(NaN)
  const alpha = 2 / (length + 1)
  let start = values.findIndex((v) => !Number.isNaN(v))
  if (start < 0) return out
  out[start] = values[start]
  for (let i = start + 1; i < values.length; i++) {
    out[i] = Number.isNaN(values[i]) ? out[i - 1] : alpha * values[i] + (1 - alpha) * out[i - 1]
  }
  return out
}

export function computeHilegaMilega(closes: number[], params: IndicatorParams) {
  const rsi = computeRsi(closes, params.rsiLength)
  const hilega = computeWma(rsi, params.wmaLength)
  const milega = computeEma(rsi, params.emaLength)
  return { rsi, hilega, milega }
}

export interface SignalMarker {
  time: number
  side: 'BUY' | 'SELL'
}

/** Detect Hilega Milega crossover signals on chart bars */
export function detectSignals(
  times: number[],
  closes: number[],
  params: IndicatorParams,
): SignalMarker[] {
  const { hilega, milega } = computeHilegaMilega(closes, params)
  const signals: SignalMarker[] = []
  for (let i = 1; i < closes.length; i++) {
    if (Number.isNaN(hilega[i]) || Number.isNaN(milega[i])) continue
    const pm = milega[i - 1]
    const ph = hilega[i - 1]
    const m = milega[i]
    const h = hilega[i]
    const longCross = pm <= ph && m > h && m > params.midLevel && h > params.midLevel
    const shortCross = pm >= ph && m < h && m < params.midLevel && h < params.midLevel
    if (longCross) signals.push({ time: times[i], side: 'BUY' })
    if (shortCross) signals.push({ time: times[i], side: 'SELL' })
  }
  return signals
}

File: backend/app/engine/indicators.py
"""Aladin indicator calculations."""

from dataclasses import dataclass

import numpy as np


@dataclass
class IndicatorParams:
    rsi_length: int = 14
    wma_length: int = 21
    ema_length: int = 3
    mid_level: float = 50.0


@dataclass
class IndicatorSnapshot:
    rsi: float
    aladin_signal: float  # WMA of RSI
    aladin_fast: float  # EMA of RSI
    long_bias: bool
    short_bias: bool
    long_cross: bool
    short_cross: bool

    @property
    def hilega(self) -> float:
        return self.aladin_signal

    @property
    def milega(self) -> float:
        return self.aladin_fast


def compute_rsi(closes: np.ndarray, length: int = 14) -> np.ndarray:
    if len(closes) < length + 1:
        return np.full(len(closes), np.nan)
    deltas = np.diff(closes)
    gains = np.where(deltas > 0, deltas, 0.0)
    losses = np.where(deltas < 0, -deltas, 0.0)
    rsi = np.full(len(closes), np.nan)
    avg_gain = np.mean(gains[:length])
    avg_loss = np.mean(losses[:length])
    if avg_loss == 0:
        rsi[length] = 100.0
    else:
        rs = avg_gain / avg_loss
        rsi[length] = 100.0 - (100.0 / (1.0 + rs))
    for i in range(length + 1, len(closes)):
        avg_gain = (avg_gain * (length - 1) + gains[i - 1]) / length
        avg_loss = (avg_loss * (length - 1) + losses[i - 1]) / length
        if avg_loss == 0:
            rsi[i] = 100.0
        else:
            rs = avg_gain / avg_loss
            rsi[i] = 100.0 - (100.0 / (1.0 + rs))
    return rsi


def wma(values: np.ndarray, length: int) -> np.ndarray:
    result = np.full(len(values), np.nan)
    weights = np.arange(1, length + 1, dtype=float)
    weight_sum = weights.sum()
    for i in range(length - 1, len(values)):
        window = values[i - length + 1 : i + 1]
        if np.any(np.isnan(window)):
            continue
        result[i] = np.dot(window, weights) / weight_sum
    return result


def ema(values: np.ndarray, length: int) -> np.ndarray:
    result = np.full(len(values), np.nan)
    alpha = 2.0 / (length + 1)
    start = 0
    while start < len(values) and np.isnan(values[start]):
        start += 1
    if start >= len(values):
        return result
    result[start] = values[start]
    for i in range(start + 1, len(values)):
        if np.isnan(values[i]):
            result[i] = result[i - 1]
        else:
            result[i] = alpha * values[i] + (1 - alpha) * result[i - 1]
    return result


def compute_aladin(
    closes: np.ndarray, params: IndicatorParams | None = None
) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
    params = params or IndicatorParams()
    rsi = compute_rsi(closes, params.rsi_length)
    aladin_signal = wma(rsi, params.wma_length)
    aladin_fast = ema(rsi, params.ema_length)
    return rsi, aladin_signal, aladin_fast


def compute_hilega_milega(
    closes: np.ndarray, params: IndicatorParams | None = None
) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
    """Backward-compatible alias."""
    return compute_aladin(closes, params)


def snapshot_at(
    closes: np.ndarray, params: IndicatorParams | None = None
) -> IndicatorSnapshot | None:
    if len(closes) < 30:
        return None
    rsi_arr, signal_arr, fast_arr = compute_aladin(closes, params)
    params = params or IndicatorParams()
    i = len(closes) - 1
    prev = i - 1
    if np.isnan(rsi_arr[i]) or np.isnan(signal_arr[i]) or np.isnan(fast_arr[i]):
        return None
    if prev < 0 or np.isnan(fast_arr[prev]) or np.isnan(signal_arr[prev]):
        return None

    aladin_fast, aladin_signal = float(fast_arr[i]), float(signal_arr[i])
    prev_fast, prev_signal = float(fast_arr[prev]), float(signal_arr[prev])
    mid = params.mid_level

    long_cross = prev_fast <= prev_signal and aladin_fast > aladin_signal
    short_cross = prev_fast >= prev_signal and aladin_fast < aladin_signal
    long_bias = long_cross and aladin_fast > mid and aladin_signal > mid
    short_bias = short_cross and aladin_fast < mid and aladin_signal < mid

    return IndicatorSnapshot(
        rsi=float(rsi_arr[i]),
        aladin_signal=aladin_signal,
        aladin_fast=aladin_fast,
        long_bias=long_bias,
        short_bias=short_bias,
        long_cross=long_cross,
        short_cross=short_cross,
    )


def compute_atr(highs: np.ndarray, lows: np.ndarray, closes: np.ndarray, length: int = 14) -> float:
    if len(closes) < length + 1:
        return 0.0
    trs = []
    for i in range(1, len(closes)):
        tr = max(
            highs[i] - lows[i],
            abs(highs[i] - closes[i - 1]),
            abs(lows[i] - closes[i - 1]),
        )
        trs.append(tr)
    return float(np.mean(trs[-length:]))
